includes/class-wc-subscription-data-copier.php

```diff
@@ -0,0 +1,463 @@
+<?php
+/**
+ * Woocommerce Subscriptions Data Copier
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+class WC_Subscription_Data_Copier {
+
+	/**
+	 * The default copy type.
+	 */
+	const DEFAULT_COPY_TYPE = 'subscription';
+
+	/**
+	 * The data keys that are excluded from the copy.
+	 *
+	 * @var string[]
+	 */
+	const DEFAULT_EXCLUDED_META_KEYS = array(
+		'_paid_date',
+		'_date_paid',
+		'_completed_date',
+		'_date_completed',
+		'_edit_last',
+		'_subscription_switch_data',
+		'_order_key',
+		'_edit_lock',
+		'_wc_points_earned',
+		'_transaction_id',
+		'_billing_interval',
+		'_billing_period',
+		'_subscription_resubscribe',
+		'_subscription_renewal',
+		'_subscription_switch',
+		'_payment_method',
+		'_payment_method_title',
+		'_suspension_count',
+		'_requires_manual_renewal',
+		'_cancelled_email_sent',
+		'_trial_period',
+		'_created_via',
+		'_order_stock_reduced',
+	);
+
+	/**
+	 * The subscription being copied.
+	 *
+	 * @var WC_Order
+	 */
+	private $from_object = null;
+
+	/**
+	 * The subscription being copied to.
+	 *
+	 * @var WC_Order
+	 */
+	private $to_object = null;
+
+	/**
+	 * The type of copy. Can be 'subscription' or 'renewal'. Optional. Default is 'subscription'.
+	 *
+	 * Used in dynamic filters to allow third parties to target specific meta keys in different copying contexts.
+	 *
+	 * @var string
+	 */
+	private $copy_type = '';
+
+	/**
+	 * Gets an instance of a copier.
+	 *
+	 * @param WC_Order $from_object The object to copy data from.
+	 * @param WC_Order $to_object   The object to copy data to.
+	 * @param string                   $copy_type   The type of copy. Can be 'subscription', 'parent', 'renewal_order' or 'resubscribe_order'. Optional. Default is 'subscription'.
+	 *
+	 * @return WC_Subscription_Data_Copier
+	 */
+	public static function get_instance( $from_object, $to_object, $copy_type = self::DEFAULT_COPY_TYPE ) {
```

Comment:
```suggestion
	public static function instance( $from_object, $to_object, $copy_type = self::DEFAULT_COPY_TYPE ) {
```

Thoughts on renaming this to just `instance()`?

This makes it consistent with the similar function we use for `WC_Subscriptions_Plugin_Core::instance()` and also is similar to what WC Core does.

Another option I was thinking of was naming this public static function "copy" and have it automatically call `this->copy_data()`

`WC_Subscription_Data_Copier::copy( $from, $to, subscription );`

Reason: The only reason we use this copier class is to copy data so I don't see much use cases where we need to get an instance of the class and do something else other than copy data. So why not make the static helper function create the instance and call the copy function? 😄